File: frontend/tests/dispatch.test.mjs
// dispatch 块的归约测试：子上下文隔离（归属路由）+ 状态定格 + 主时间线不受污染。
import test from 'node:test';
import assert from 'node:assert/strict';
import { reduce } from '../src/shared/store.ts';
const base = { blocks: [], busy: true, pending: null, todos: [], currentId: '', operationError: null };

test('dispatchStart 开卡 → 子事件按 dispatchId 挂进 subBlocks', () => {
  let s = reduce(base, { type: 'dispatchStart', dispatchId: 'd1', agentId: 'coder', agentName: '代码 Agent', agentColor: '#3b82f6', task: '跑测试' });
  s = reduce(s, { type: 'delta', kind: 'text', text: '子回复', dispatchId: 'd1' });
  s = reduce(s, { type: 'toolCall', id: 'd-c1', name: 'bash', arguments: '{}', dispatchId: 'd1' });
  s = reduce(s, { type: 'toolResult', id: 'd-c1', name: 'bash', content: 'PASS', isError: false, dispatchId: 'd1' });
  const card = s.blocks.find((b) => b.kind === 'dispatch');
  assert.equal(card.kind, 'dispatch');
  assert.equal(card.status, 'running');
  assert.equal(card.subBlocks.length, 2, '子块两个（assistant + tool）');
  assert.equal(card.subBlocks[0].content, '子回复');
  assert.equal(card.subBlocks[1].result, 'PASS');
  // 主时间线没有子内容（隔离）
  assert.equal(s.blocks.filter((b) => b.kind === 'assistant').length, 0);
  assert.equal(s.blocks.filter((b) => b.kind === 'tool').length, 0);
});

test('dispatchEnd 定格 + 结果回填（done 状态 + result）', () => {
  let s = reduce(base, { type: 'dispatchStart', dispatchId: 'd1', agentId: 'coder', agentName: '代码', agentColor: '#123456', task: 't' });
  s = reduce(s, { type: 'delta', kind: 'text', text: '部分', dispatchId: 'd1' });
  s = reduce(s, { type: 'done', usageTokens: 100, finishReason: 'stop', dispatchId: 'd1' });
  s = reduce(s, { type: 'dispatchEnd', dispatchId: 'd1', result: '最终结论', isError: false, usageTokens: 100 });
  const card = s.blocks.find((b) => b.kind === 'dispatch');
  assert.equal(card.status, 'done');
  assert.equal(card.result, '最终结论');
  // 子时间线的 assistant 已定格带 usage
  const sub = card.subBlocks.find((b) => b.kind === 'assistant');
  assert.equal(sub.streaming, false);
  assert.equal(sub.usageTokens, 100);
});

test('未知 dispatchId 的子事件被忽略（不炸不挂错）', () => {
  const s = reduce(base, { type: 'delta', kind: 'text', text: '孤儿', dispatchId: 'nobody' });
  assert.deepEqual(s.blocks, []);
});

test('子事件不干扰主时间线的流式块（并行场景的归属正确性）', () => {
  let s = reduce(base, { type: 'dispatchStart', dispatchId: 'd1', agentId: 'coder', agentName: 'c', agentColor: '#000', task: 't' });
  s = reduce(s, { type: 'delta', kind: 'text', text: '主时间线' });
  s = reduce(s, { type: 'delta', kind: 'text', text: '子时间线', dispatchId: 'd1' });
  const main = s.blocks.find((b) => b.kind === 'assistant');
  const card = s.blocks.find((b) => b.kind === 'dispatch');
  assert.equal(main.content, '主时间线');
  assert.equal(card.subBlocks[0].content, '子时间线');
});

File: internal/agent/dispatch_test.go
// M3 agent.dispatch 的内核测试：主 Agent 派发 → 子上下文隔离执行 →
// 结果回填主历史；委派名单校验；事件归属标记。
package agent

import (
	"context"
	"encoding/json"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/moyunteng/lxcode/internal/config"
	"github.com/moyunteng/lxcode/internal/llm"
	"github.com/moyunteng/lxcode/internal/sessiondata"
	"github.com/moyunteng/lxcode/internal/tools"
)

// dispatchEnv 是 dispatchSession 的可观测面（事件捕获）。
type dispatchEnv struct {
	s      *Session
	events []Event
	mu     sync.Mutex
}

func (e *dispatchEnv) snapshot() []Event {
	e.mu.Lock()
	defer e.mu.Unlock()
	return append([]Event(nil), e.events...)
}

// newDispatchSession 构造挂 resolver 的会话（main 委派 coder）+ 可编程假流。
func newDispatchSession(t *testing.T, stream StreamFn) *dispatchEnv {
	t.Helper()
	reg, err := config.Load(filepath.Join(t.TempDir(), "config", "models.json"))
	if err != nil {
		t.Fatal(err)
	}
	if err := reg.Add(config.ModelConfig{
		ID: "m1", BaseURL: "http://127.0.0.1:1/v1", Model: "m1", Enabled: true,
		Capabilities: config.Capabilities{Tools: true},
	}); err != nil {
		t.Fatal(err)
	}
	if err := reg.SetRole(config.RoleDefault, "m1"); err != nil {
		t.Fatal(err)
	}
	env := &dispatchEnv{}
	s := New(reg, tools.New(), func(ev Event) {
		env.mu.Lock()
		env.events = append(env.events, ev)
		env.mu.Unlock()
	})
	s.SetStream(stream)
	t.Cleanup(s.Close)
	s.SetAgentResolver(&stubResolver{entries: map[string]*sessiondata.AgentContext{
		"main": {
			Def: sessiondata.AgentDef{ID: "main", Name: "主 Agent", IsMain: true, Enabled: true,
				Tools: []string{"agent.dispatch"}, Delegates: []string{"coder"}},
			Delegates: []sessiondata.AgentDef{
				{ID: "coder", Name: "代码 Agent", Desc: "写代码", Enabled: true},
			},
		},
		"coder": {
			Def: sessiondata.AgentDef{ID: "coder", Name: "代码 Agent", Enabled: true, Color: "#3b82f6",
				Tools: []string{"read_file", "edit"}, Approval: "confirm"},
		},
		"outsider": {
			Def: sessiondata.AgentDef{ID: "outsider", Name: "名单外", Enabled: true, Tools: []string{}},
		},
	}})
	env.s = s
	return env
}

// waitDispatchIdle 等会话空闲（轮跑完）。
func waitDispatchIdle(t *testing.T, s *Session) {
	t.Helper()
	for i := 0; i < 300 && s.Busy(); i++ {
		time.Sleep(10 * time.Millisecond)
	}
}

func TestDispatchSubContextIsolation(t *testing.T) {
	var mainCalls, subCalls int
	var subPrompt, subFirstMsg string
	env := newDispatchSession(t, func(ctx context.Context, m config.ModelConfig, msgs []llm.Message, opts []llm.Option) (<-chan llm.StreamEvent, error) {
		ch := make(chan llm.StreamEvent, 4)
		if strings.Contains(msgs[0].Content, "主 Agent（调度中枢）") {
			// 主轮：第一轮发起 dispatch；拿到工具结果后收尾（模拟验收汇总）
			mainCalls++
			if mainCalls == 1 {
				tc := llm.ToolCall{ID: "call-1"}
				tc.Function.Name = "agent.dispatch"
				tc.Function.Arguments = `{"agent":"coder","task":"读 README 并总结","context":"项目根在当前目录"}`
				go func() {
					defer close(ch)
					ch <- llm.StreamEvent{Type: llm.EventToolCall, ToolCall: tc}
					ch <- llm.StreamEvent{Type: llm.EventDone, Result: &llm.ChatResult{
						Message: llm.Message{Role: "assistant", ToolCalls: []llm.ToolCall{tc}}, FinishReason: llm.FinishToolCalls,
					}}
				}()
				return ch, nil
			}
			go func() {
				defer close(ch)
				ch <- llm.StreamEvent{Type: llm.EventDone, Result: &llm.ChatResult{
					Message: llm.Message{Role: "assistant", Content: "已验收。"}, FinishReason: llm.FinishStop,
				}}
			}()
			return ch, nil
		}
		// 子轮
		subCalls++
		subPrompt = msgs[0].Content
		subFirstMsg = msgs[1].Content
		if subCalls == 1 {
			tc := llm.ToolCall{ID: "sub-1"}
			tc.Function.Name = "read_file"
			tc.Function.Arguments = `{"path":"README.md"}`
			go func() {
				defer close(ch)
				ch <- llm.StreamEvent{Type: llm.EventToolCall, ToolCall: tc}
				ch <- llm.StreamEvent{Type: llm.EventDone, Result: &llm.ChatResult{
					Message: llm.Message{Role: "assistant", ToolCalls: []llm.ToolCall{tc}}, FinishReason: llm.FinishToolCalls,
				}}
			}()
			return ch, nil
		}
		go func() {
			defer close(ch)
			ch <- llm.StreamEvent{Type: llm.EventText, TextDelta: "README 总结完成：这是一个 agent 项目。"}
			ch <- llm.StreamEvent{Type: llm.EventDone, Result: &llm.ChatResult{
				Message: llm.Message{Role: "assistant", Content: "README 总结完成：这是一个 agent 项目。"}, FinishReason: llm.FinishStop,
			}}
		}()
		return ch, nil
	})

	if err := env.s.Send("帮我总结 README"); err != nil {
		t.Fatal(err)
	}
	waitDispatchIdle(t, env.s)

	if mainCalls != 2 || subCalls != 2 {
		t.Fatalf("调用数不符 main=%d sub=%d", mainCalls, subCalls)
	}
	// 隔离验证：子上下文第一条 = 任务消息（不是主历史）
	if !strings.Contains(subFirstMsg, "读 README 并总结") || !strings.Contains(subFirstMsg, "项目根在当前目录") {
		t.Fatalf("子上下文应只含任务+背景: %q", subFirstMsg)
	}
	if strings.Contains(subFirstMsg, "帮我总结 README") {
		t.Fatalf("子上下文不应含主历史: %q", subFirstMsg)
	}
	// 子提示词 = 执行协议（子 Agent 四层组合）
	if !strings.Contains(subPrompt, "你是执行 Agent") {
		t.Fatal("子提示词应为执行协议")
	}
	// 主历史：user + dispatch 工具结果（子结论回填）——子中间消息不进主历史
	var toolResultMsg string
	var subToolMsgs int
	for _, m := range env.s.History().Messages {
		if m.Role == "tool" {
			toolResultMsg = m.Content
			subToolMsgs++
		}
	}
	if subToolMsgs != 1 {
		t.Fatalf("主历史应只有一条工具结果（子中间消息不进主历史）: %d", subToolMsgs)
	}
	if !strings.Contains(toolResultMsg, "README 总结完成") {
		t.Fatalf("子结论应回填主历史（工具结果）: %q", toolResultMsg)
	}
	// 事件序：DispatchStart → 子 delta（带 DispatchID=call-1）→ DispatchEnd
	var sawStart, sawSubDelta, sawSubDone, sawEnd bool
	for _, ev := range env.snapshot() {
		if e, ok := ev.(DispatchStartEvent); ok {
			sawStart = e.DispatchID == "call-1" && e.AgentID == "coder" && e.AgentColor != ""
		}
		if e, ok := ev.(DeltaEvent); ok && e.DispatchID == "call-1" {
			sawSubDelta = strings.Contains(e.Text, "README 总结完成")
		}
		if e, ok := ev.(TurnDoneEvent); ok && e.DispatchID == "call-1" {
			sawSubDone = true
		}
		if e, ok := ev.(DispatchEndEvent); ok {
			sawEnd = e.DispatchID == "call-1" && !e.IsError && strings.Contains(e.Result, "README 总结完成")
		}
	}
	if !sawStart || !sawSubDelta || !sawSubDone || !sawEnd {
		t.Fatalf("事件序缺失 start=%v subDelta=%v subDone=%v end=%v", sawStart, sawSubDelta, sawSubDone, sawEnd)
	}
}

func TestDispatchGuards(t *testing.T) {
	run := func(t *testing.T, target string, wantErr string) {
		t.Helper()
		var result string
		env := newDispatchSession(t, func(ctx context.Context, m config.ModelConfig, msgs []llm.Message, opts []llm.Option) (<-chan llm.StreamEvent, error) {
			ch := make(chan llm.StreamEvent, 4)
			if strings.Contains(msgs[0].Content, "调度中枢") {
				tc := llm.ToolCall{ID: "c1"}
				tc.Function.Name = "agent.dispatch"
				tc.Function.Arguments = `{"agent":"` + target + `","task":"x"}`
				go func() {
					defer close(ch)
					ch <- llm.StreamEvent{Type: llm.EventToolCall, ToolCall: tc}
					ch <- llm.StreamEvent{Type: llm.EventDone, Result: &llm.ChatResult{
						Message: llm.Message{Role: "assistant", ToolCalls: []llm.ToolCall{tc}}, FinishReason: llm.FinishToolCalls}}
				}()
				return ch, nil
			}
			go func() {
				defer close(ch)
				ch <- llm.StreamEvent{Type: llm.EventDone, Result: &llm.ChatResult{
					Message: llm.Message{Role: "assistant", Content: "ok"}}}
			}()
			return ch, nil
		})
		if err := env.s.Send("派活"); err != nil {
			t.Fatal(err)
		}
		waitDispatchIdle(t, env.s)
		for _, m := range env.s.History().Messages {
			if m.Role == "tool" {
				result = m.Content
			}
		}
		if !strings.Contains(result, wantErr) {
			t.Fatalf("目标 %s 应拒绝（%s）: %q", target, wantErr, result)
		}
	}
	t.Run("名单外拒绝", func(t *testing.T) { run(t, "outsider", "不在有效委派名单内") })
	t.Run("主 Agent 不可被委派", func(t *testing.T) { run(t, "main", "主 Agent 不可被委派") })
	t.Run("不存在的 Agent", func(t *testing.T) { run(t, "nobody", "Agent 不存在") })
}

// JSON 参数解析路径（runTools 调用位的展开形状）。
func TestDispatchCallUnmarshal(t *testing.T) {
	var p struct {
		Agent   string `json:"agent"`
		Task    string `json:"task"`
		Context string `json:"context"`
	}
	if err := json.Unmarshal([]byte(`{"agent":"coder","task":"t","context":"c"}`), &p); err != nil {
		t.Fatal(err)
	}
	if p.Agent != "coder" || p.Task != "t" || p.Context != "c" {
		t.Fatalf("解析失真: %+v", p)
	}
}
